api/src/main/java/org/apache/iceberg/types/Types.java

```diff
@@ -694,7 +698,10 @@ private static Literal<?> castDefault(Literal<?> defaultValue, Type type) {
         throw new IllegalArgumentException(
             String.format("Invalid default value for %s: %s (must be null)", type, defaultValue));
       } else if (defaultValue != null) {
-        return defaultValue.to(type);
+        Literal<?> typedDefault = defaultValue.to(type);
```

Comment:
Thinking of it, would a `NoneLiteral` make sense 🤔

Reply:
Right now, we make a lot of assumptions that Literal does not contain null or NaN. That way, we don't need to implement 3-value boolean logic or worry about how to handle NaN in comparisons.

If you want to have a null default, then you don't set a default value for an optional column. And this also passes through null without a problem if the Literal was null.